drivers/char/tpm/tpm_ftpm_optee.c

```diff
@@ -0,0 +1,304 @@
+
+	shm = tee_shm_alloc(
+		tpm_tee_context.ftpm_optee_context, 
+		len + sizeof(tpm_tee_context.response_buffer), 
+		TEE_SHM_MAPPED | TEE_SHM_DMA_BUF);
+
+	if (IS_ERR(shm))
+	{
+		rc = -ENOMEM;
+		goto cleanup;
+	}
+
+	// request
+	submit_command_params[0].u.memref.size = len;
+	submit_command_params[0].u.memref.shm = shm;
+	submit_command_params[0].u.memref.shm_offs = 0;
+	memcpy(submit_command_params[0].u.memref.shm->kaddr, buf, len);
+
+	// response
+	submit_command_params[1].u.memref.size = sizeof(tpm_tee_context.response_buffer);
+	submit_command_params[1].u.memref.shm = shm;
+	submit_command_params[1].u.memref.shm_offs = len;
+
+	rc = tee_client_invoke_func(
+		tpm_tee_context.ftpm_optee_context, 
+		&submit_command_args, 
+		submit_command_params);
+	if (rc || submit_command_args.ret)
+	{
+		rc = rc ? rc : submit_command_args.ret;
+		goto cleanup;
+	}
+
+	response_buffer = submit_command_params[1].u.memref.shm->kaddr + 
+		submit_command_params[1].u.memref.shm_offs;
+	response_header = (struct tpm_output_header*)response_buffer;
+	response_length = be32_to_cpu(response_header->length);
+
+	if (response_length > sizeof(tpm_tee_context.response_buffer))
+	{
+		rc = -EIO;
+		goto cleanup;
+	}
+
+	memcpy(
+		tpm_tee_context.response_buffer, 
+		response_buffer, 
+		response_length);
+	tpm_tee_context.response_length = response_length;
+
+cleanup:
+	if (shm)
+	{
+		tee_shm_free(shm);
+	}
+
+	mutex_unlock(&tpm_tee_context.response_buffer_lock);
+
+	return rc;
+}
+
+static void ftpm_optee_tpm_op_cancel(struct tpm_chip *chip)
+{
+	;
+}
+
+static u8 ftpm_optee_tpm_op_status(struct tpm_chip *chip)
+{
+	return 0;
+}
+
+static bool ftpm_optee_tpm_req_canceled(struct tpm_chip *chip, u8 status)
+{
+	return 0;
+}
+
+static const struct tpm_class_ops ftpm_optee_tpm_ops = {
+	.flags = TPM_OPS_AUTO_STARTUP,
+	.req_complete_mask = 0,
+	.req_complete_val = 0,
+	.recv = ftpm_optee_tpm_op_recv,
+	.send = ftpm_optee_tpm_op_send,
+	.cancel = ftpm_optee_tpm_op_cancel,
+	.status = ftpm_optee_tpm_op_status,
+	.req_canceled = ftpm_optee_tpm_req_canceled
+};
+
+static void ftpm_optee_cleanup(void)
+{
+	if (tpm_tee_context.chip_registered)
+		tpm_chip_unregister(tpm_tee_context.chip);
+
+	if (tpm_tee_context.chip != NULL)
+		put_device(&tpm_tee_context.chip->dev);
+
+	if (tpm_tee_context.ftpm_optee_context != NULL)
+	{
+		if (tpm_tee_context.optee_session_opened)
+		{
+			tee_client_close_session(
+				tpm_tee_context.ftpm_optee_context, 
+				tpm_tee_context.optee_session_id);
+			tpm_tee_context.optee_session_opened = FALSE;
+		}
+
+		tee_client_close_context(tpm_tee_context.ftpm_optee_context);
+		tpm_tee_context.ftpm_optee_context = NULL;
+	}
+
+	mutex_destroy(&tpm_tee_context.response_buffer_lock);
+}
+
+static int ftpm_opteee_match_func(
+	struct tee_ioctl_version_data *version,
+	const void * data)
+{
+	if (version->impl_id == TEE_OPTEE_CAP_TZ &&
+		version->impl_caps == TEE_IMPL_ID_OPTEE &&
+		version->gen_caps == TEE_GEN_CAP_GP)
```

Comment:
Is the extra indentation on purpose?

Reply:
\t, kernel seem to be using that for indentation